File: scripts/gradient_descent.jl
using LinearAlgebra

function gradient_descent(P, q, x₀; α=0.1,
                maxiter=1000, ϵ=1e-5)
    x = copy(x₀)
    ∇f = x -> P * x + q
    Δx = -∇f(x)
    iter = 0
    while norm(Δx) > ϵ || iter ≤ maxiter
        iter += 1
        x .+= α * Δx
        Δx .= - ∇f(x)
    end
    return x
end

P = [10.0 -1.0;
    -1.0 1.0];
q = [0; -10.0];
x₀ = zeros(2);

File: scripts/gradient_descent.py
import numpy as np

def gradient_descent(P, q, x0, alpha=0.1,
                maxiter=1000, epsilon=1e-5):
    x = np.copy(x0)
    grad_x = P.dot(x) + q
    iter = 0
    while np.linalg.norm(grad_x) > epsilon or\
               iter <= maxiter:
        iter += 1
        x -= alpha * grad_x
        grad_x[:] = P.dot(x) + q
    return x

P = np.array([[10, -1],
              [-1, 1]])
q = np.array([[0], [-10]])
x0 = np.zeros((2, 1))

gradient_descent(P, q, x0)
